### PR TITLE
feat: multiple database backends on separate paths

### PR DESCRIPTION
**Pull request by gkanwar**
Config file specified by the `OMNIBOARD_CONFIG` environment variable defines the set of paths, according to the discussion in issue #69 

Outstanding TODOs:

- [x] ~Improvements to config file format~
- [x] Resolve frontend quirks like local storage per context path
- [x] Link each context path in the top bar
- [x] Ensure web tests pass

### CODE REVIEW

**Comment by vivekratnavel** [comment]
I can help you out on this. When I tried your changes in my local environment, I get a 404 error while fetching `/api/v1/databases`.

I tried with this config:
```
{
  "default": {
    "mongodbURI": "mongodb://dockerhost:27019/sacred",
    "path":"/"
  },
  "fresh": {
	"mongodbURI": "mongodb://dockerhost:27020/sacred",
    "path":"/fresh"
  }
} 
```

What is working and not working in the current state?

**Comment by gkanwar** [comment]
Thanks! Ah, I think the root path '/' needs to be reserved: '/api/v1/databases' is handled outside the router for a specific backend. I'll make a fix later today.

If you idea for how to separate out the local storage for each context that would be useful to know.

**Comment by gkanwar** [comment]
I added error checking for root path in 2c84351, and fixed a minor navbar key issue. I don't think we need any enhancements to the database config format before landing this, so striking that item. That leaves just splitting localStorage to be scoped to each context independently.

**Comment by vivekratnavel** [comment]
I have an idea to handle localstorage. 

We can use `this.props.location.pathname` to get the path and use it as a prefix for localstorage keys.

Example: Before getting or setting localstorage, we can get pathname and use it as prefix for DEFAULT_SELECTION_KEY - https://github.com/vivekratnavel/omniboard/blob/b18a2a2a3ea4103b1b0a8da2c582982d161e2f4c/web/src/components/MetricsPlotView/metricsPlotView.js#L13

This [line](https://github.com/vivekratnavel/omniboard/blob/b18a2a2a3ea4103b1b0a8da2c582982d161e2f4c/web/src/components/MetricsPlotView/metricsPlotView.js#L13) would be 
```
const selectionKey = this.props.location.pathname + "|" + DEFAULT_SELECTION_KEY;
const defaultSelection = JSON.parse(localStorage.getItem(selectionKey))
```

**Comment by gkanwar** [comment]
Thanks. How do you feel about using the database key (the object key in the config file) as this prefix instead of path. Using path feels a little brittle and passing a key around is not too hard.

**Comment by vivekratnavel** [comment]
Sure, that sounds good to me. The only advantage with path might be that, you can retain the settings even if you change the database path. But, I am fine with any unique prefix.

**Comment by gkanwar** [comment]
I had to add a small loading interstitial while waiting for a response from /api/v1/database, however this seems to break many of the tests since that response is async. Any idea how to wait for this in the tests?

**Comment by gkanwar** [comment]
Okay, tests resolved. I think this is ready for a full review.

**Comment by vivekratnavel** [comment]
With the latest changes, I get the following error :
```
Error: No default engine was specified and no extension was provided.
    at new View (/Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/view.js:61:11)
    at Function.render (/Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/application.js:570:12)
    at ServerResponse.render (/Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/response.js:1012:7)
    at errorHandler (/Users/vivekratnavel/Documents/github-omniboard/server/index.js:70:7)
    at Layer.handle_error (/Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/router/layer.js:71:5)
    at trim_prefix (/Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/router/index.js:315:13)
    at /Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/router/index.js:335:12)
    at next (/Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/router/index.js:275:10)
    at clientErrorHandler (/Users/vivekratnavel/Documents/github-omniboard/server/index.js:64:5)
    at Layer.handle_error (/Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/router/layer.js:71:5)
    at trim_prefix (/Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/router/index.js:315:13)
    at /Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/router/index.js:335:12)
    at next (/Users/vivekratnavel/Documents/github-omniboard/node_modules/express/lib/router/index.js:275:10)
    at logErrors (/Users/vivekratnavel/Documents/github-omniboard/server/index.js:57:3)
```

File: server/index.js [hunk 1]
```diff
@@ -1,6 +1,74 @@
...
+  if (req.xhr) {
+    res.status(500).send({ message: 'An unknown error occurred!' });
+  } else {
+    next(err);
+  }
+}
+
+function errorHandler (err, req, res, next) {
+  res.status(500);
+  res.render('error', { message: err });
```
Comment (vivekratnavel) [inline server/index.js]:
```suggestion
  res.send({ message: err.message });
```

**Comment by vivekratnavel** [comment]
Able to get it working. Looks nice! Let me play around with it for a while and post a full review of the PR. Thanks!

**Comment by vivekratnavel** [comment]
I am unable to view / download source files. Can you confirm if it's a bug?

It might be because of this - https://github.com/vivekratnavel/omniboard/pull/200/files#diff-7b82b1a4b8e4beda79facc67e81748f3R55

**Comment by vivekratnavel** [comment]
There are two issues I see:

- Even though axios call is relative in sourceFilesView, the actual network call being triggered is not relative
- For the API call <path>/api/v1/files/preview/<file_id> , the response from server is an error 
 {
"message": "Cannot read property 'createReadStream' of null"
}
The error is caused because of this - https://github.com/vivekratnavel/omniboard/pull/200/files#diff-7b82b1a4b8e4beda79facc67e81748f3R55
dbExport will always be returned before `dbExport.gfs = Grid(db.db, mongoose.mongo)` is executed since db.once is async.

File: web/src/components/App/index.js [hunk 2]
```diff
@@ -164,8 +186,10 @@ class App extends Component {
             </Navbar.Brand>
           </Navbar.Header>
           <Navbar.Collapse>
-            <Nav pullLeft>
-              <NavItem>({dbName})</NavItem>
+            <Nav pullLeft activeKey={dbInfo.key}>
+              {otherDbs.map(db => {
+                return <NavItem key={db.key} eventKey={db.key} href={db.path}>({db.name})</NavItem>;
```
Comment (vivekratnavel) [inline web/src/components/App/index.js]: Let's also add key to the navbar.
```suggestion
                return <NavItem key={db.key} eventKey={db.key} href={db.path}>{db.key} ({db.name})</NavItem>;
```

**Comment by vivekratnavel** [comment]
Can you also add documentation for this feature with sample json file? 

Thanks!

**Comment by gkanwar** [comment]
Resolved the source files issue by threading database info through the various components so api calls can be appropriately scoped. Added docs and removed the redundant sample json file.

**Comment by gkanwar** [comment]
@vivekratnavel anything else left to do for this PR?

**Comment by vivekratnavel** [comment]
@nuth & @gkanwar Thanks for your contribution! I have merged the PR and decided to track and fix any issues as bugs in the future.

**Pull request merged**